bigtable/internal/testproxy/proxy.go

```diff
@@ -0,0 +1,963 @@
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to CheckAndMutateRow() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+
+	ctx = btc.addCancelFunction(ctx)
+
+	keys, err := t.SampleRowKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sk := make([]*btpb.SampleRowKeysResponse, 0)
+	for _, k := range keys {
+		s := &btpb.SampleRowKeysResponse{
+			RowKey: []byte(k),
+		}
+		sk = append(sk, s)
+	}
+
+	res := &pb.SampleRowKeysResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Sample: sk,
+	}
+	return res, nil
+}
+
+// ReadModifyWriteRow responds to the ReadModifyWriteRow RPC. This method
+// applies a non-idempotent change to a row.
+func (s *goTestProxyServer) ReadModifyWriteRow(ctx context.Context, req *pb.ReadModifyWriteRowRequest) (*pb.RowResult, error) {
+	btc, exists := s.clientIDs[req.ClientId]
+	if !exists {
+		return nil, stat.Error(codes.InvalidArgument,
+			fmt.Sprintf("%s: ClientID does not exist", logLabel))
+	}
+
+	rrq := req.GetRequest()
+	if rrq == nil {
+		return nil, stat.Error(codes.InvalidArgument, "request to CheckAndMutateRow() is missing inner request")
+	}
+
+	t := btc.c.Open(rrq.TableName)
+	k := string(rrq.RowKey)
+	rpb := rrq.Rules
+
+	rmw := bigtable.NewReadModifyWrite()
+
+	for _, rp := range rpb {
+		switch r := rp.Rule; r.(type) {
+		case *btpb.ReadModifyWriteRule_AppendValue:
+			av := r.(*btpb.ReadModifyWriteRule_AppendValue)
+			rmw.AppendValue(rp.FamilyName, string(rp.ColumnQualifier), av.AppendValue)
+		case *btpb.ReadModifyWriteRule_IncrementAmount:
+			ia := r.(*btpb.ReadModifyWriteRule_IncrementAmount)
+			rmw.Increment(rp.FamilyName, string(rp.ColumnQualifier), ia.IncrementAmount)
+		}
+	}
+
+	var cancel context.CancelFunc
+	if btc.perOperationTimeout.AsDuration() > 0 {
+		ctx, cancel = context.WithTimeout(ctx, btc.perOperationTimeout.AsDuration())
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+	btc.cancels = append(btc.cancels, cancel)
+
+	r, err := t.ApplyReadModifyWrite(ctx, k, rmw)
+
+	if err != nil {
+		return nil, err
+	}
+
+	rp, err := rowToProto(r)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &pb.RowResult{
+		Status: &status.Status{
+			Code: int32(codes.OK),
+		},
+		Row: rp,
+	}
+	return res, nil
+}
+
+func (s *goTestProxyServer) mustEmbedUnimplementedCloudBigtableV2TestProxyServer() {}
+
+func newProxyServer(lis net.Listener) *grpc.Server {
+	s := grpc.NewServer()
+	pb.RegisterCloudBigtableV2TestProxyServer(s, &goTestProxyServer{})
+	log.Printf("server listening at %v", lis.Addr())
+	return s
+}
+
+func main() {
+	flag.Parse()
+
+	log.Println(*port)
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *port))
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
```

Comment:
this needs a `defer lis.Close()` after it

Reply:
ditto below for the proxy server.

Reply:
Done.